src/Microsoft.AspNetCore.Routing/Matchers/DfaMatcher.cs

```diff
@@ -48,115 +46,34 @@ public sealed override Task MatchAsync(HttpContext httpContext, IEndpointFeature
 
             // SelectCandidates will process the DFA and return a candidate set. This does
             // some preliminary matching of the URL (mostly the literal segments).
-            var candidates = SelectCandidates(path, segments);
-            if (candidates.GroupCount == 0)
+            var candidatesArray = FindCandidates(httpContext, path, segments);
+            if (candidatesArray.Length == 0)
             {
                 return Task.CompletedTask;
             }
 
-            // At this point we have a candidate set, defined as a list of groups
-            // of candidates. Each member of a given group has the same priority
-            // (priority is defined by order, precedence and other factors like http method
-            // or version).
+            // At this point we have a candidate set, defined as a list of endpoints in
+            // priority order.
             //
             // We don't yet know that any candidate can be considered a match, because
             // we haven't processed things like route constraints and complex segments.
             //
-            // Now we'll go group by group to capture route values, process constraints,
+            // Now we'll iterate each endpoint to capture route values, process constraints,
             // and process complex segments.
-            //
-            // Perf: using groups.Length - 1 here to elide the bounds check. We're relying
-            // on assumptions of how Groups works.
-            var candidatesArray = candidates.Candidates;
-            var groups = candidates.Groups;
-
-            for (var i = 0; i < groups.Length - 1; i++)
```

Comment:
Getting rid of 'groups' made this really simple and fast. Now if you look at a profile, the main costs are things like string comparisons and dictionary lookups, not our data structures.